action/product/list.go

```diff
@@ -26,11 +26,39 @@ type paging struct {
 // @Router /products [get]
 func list(w http.ResponseWriter, r *http.Request) {
 
-	result := paging{}
+	var nodes []productData
+	var products []model.Product
+	result := &paging{}
 
 	offset, limit := util.Paging(r.URL.Query())
 
-	model.DB.Preload("Currency").Preload("Status").Preload("ProductType").Model(&model.Product{}).Count(&result.Total).Offset(offset).Limit(limit).Find(&result.Nodes)
+	model.DB.Preload("Currency").Preload("Status").Preload("ProductType").Model(&model.Product{}).Count(&result.Total).Offset(offset).Limit(limit).Find(&products)
+
+	for _, product := range products {
+		var categories []model.ProductCategory
+		var tags []model.ProductTag
+		data := &productData{}
+		model.DB.Model(&model.ProductCategory{}).Where(&model.ProductCategory{
+			ProductID: uint(product.ID),
+		}).Preload("Category").Find(&categories)
+
+		model.DB.Model(&model.ProductTag{}).Where(&model.ProductTag{
+			ProductID: uint(product.ID),
+		}).Preload("Tag").Find(&tags)
```

Comment:
Instead of calling query inside for loop, run for loop and get all product IDs, then run one query to get ProductCategory for the array of product IDs then map them with product ID, then again run for loop to add data into product list

For reference check [dega-server space code](https://github.com/factly/dega-server/blob/feat/category/service/core/action/space/my.go#L43)